[MXNET-34] Onnx Module to import onnx models into mxnet

anirudhacharya opened the pull request:
## Description ##
Module to import onnx models into mxnet.

## Checklist ##
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage.
- [X] Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [X] Code is well-documented:
- [x] For user-facing API changes, API doc string has been updated.
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] API to import onnx models.

## Comments ##

@Roshrini @spidydev @lupesko @reminisce @piiswrong @szha
@nswamy @sandeep-krishnamurthy 
@awslabs/aws-deep-learning-sdk

reminisce commented on python/mxnet/contrib/onnx/_import/__init__.py:
```diff
@@ -0,0 +1,46 @@
+    raise ImportError("Onnx and protobuf need to be installed")
+from .import_onnx import GraphProto
+
+def import_model(model_file):
```
In MXNet, it's kind of a convention of not defining functions or classes in `__init__.py`. You can define it somewhere else and expose the API here using `import` directive.

anirudhacharya replied:
Noted. I will make changes accordingly.

anirudhacharya commented on python/mxnet/contrib/onnx/_import/__init__.py:
```diff
@@ -0,0 +1,46 @@
+    raise ImportError("Onnx and protobuf need to be installed")
+from .import_onnx import GraphProto
+
+def import_model(model_file):
```
@reminisce said - Is import_model a user level API? If so, why is it registered under a hidden module _import?

piiswrong commented on python/mxnet/contrib/onnx/_import/op_translations/negative.py:
```diff
@@ -0,0 +1,20 @@
+# under the License.
+
+# coding: utf-8
+def negative(op_name, attrs, inputs):
```
why create so many files? You can put all ops in one file

anirudhacharya replied:
As we implement more operators, it will become unwieldy to have all operator translations in the same file. The code will also be unclear. I was thinking maybe we could club similar operators into the same file like here - https://mxnet.apache.org/api/python/symbol/symbol.html#the-symbol-class

So all the arithmetic operators will be in one file, indexing routines like slice and gather in another file and so on. Will be a good balance between clubbing all the operators into the same file and having a different file for each operator.

Also some operator translations are much more involved than the ones I have made in this PR.

Roshrini replied:
@anirudhacharya Right now it seems like you are putting every single operator in a different file. (file names are also based on the operator defined).
Also, while implementing export you will have to do a reverse mapping between operators. I think that's just too many files to maintain.

+1 to have all operator conversions in a single file.

piiswrong commented on python/mxnet/contrib/onnx/_import/op_translations/_add.py:
```diff
@@ -0,0 +1,25 @@
+def _add(op_name, attrs, inputs):
+    new_attr = {}
+    if 'broadcast' in attrs and attrs['broadcast']==1:
+        return 'broadcast_add', new_attr, inputs
```
why not create the operator here directly?

anirudhacharya replied:
If I create the operator here, then the same create operator command has to be repeated for every operator translation in every method. 

Probably more cleaner and elegant to do it once in the caller method.

anirudhacharya commented:
For all the test cases to pass for this module, we will need changes in the CI process. It needs Onnx and protobuf in the CI process for the tests to run. Currently the onnx-mxnet project fetches these packages in the CI process - 
```bash
[Container] 2018/03/02 00:14:07 Running command wget https://github.com/google/protobuf/releases/download/v2.6.1/protobuf-2.6.1.tar.gz
Downloading onnx-1.0.1.tar.gz (763kB)
```

anirudhacharya commented:
@awslabs/aws-deep-learning-sdk

KellenSunderland commented:
Hi @anirudhacharya, are you blocked by the CI change?  You are able to change the CI process along with your PR, and installing two dependencies shouldn't be hard.  Just let us know if you'd like help doing this.  (To get started you can see how other python deps are installed here: https://github.com/apache/incubator-mxnet/blob/master/tests/ci_build/Dockerfile.cpu and here: https://github.com/apache/incubator-mxnet/blob/master/tests/ci_build/install/install_testdeps.sh for example.)

sandeep-krishnamurthy commented on python/mxnet/contrib/onnx/_import/translation_utils/__init__.py:
```diff
@@ -0,0 +1,141 @@
+    Using "split" to perform similar operation.
+    "split" can be slower compared to "reshape".
+     This can have performance impact.
+     TODO: Remove this implementation once mxnet adds the support.
```
Related GitHub issue link?

anirudhacharya replied:
will create an issue and add it.

sandeep-krishnamurthy commented on python/mxnet/contrib/onnx/_import/op_translations/reduce_max.py:
```diff
@@ -0,0 +1,23 @@
+
+def reduce_max(op_name, attrs, inputs):
+    new_attr = _fix_attrbute_names(attrs, {'axes':'axis'})
+    return 'max', attrs, inputs
```
nit: new line

anirudhacharya commented on python/mxnet/contrib/onnx/_import/import_helper.py:
```diff
@@ -0,0 +1,29 @@
+# coding: utf-8
+# pylint: disable=invalid-name
+"""Operator attributes conversion"""
+from . import *
```
will do.

anirudhacharya commented on python/mxnet/contrib/onnx/_import/import_onnx.py:
```diff
@@ -0,0 +1,166 @@
+                name_param = 'param_{}'.format(self._num_param)
+                self._num_param += 1
+                self._params[name_param] = self._params.pop(i.name)
+                self._nodes[name_param] = symbol.Variable(name=name_param,
```
okay. will make the change.

anirudhacharya replied:
The param values have the dtype information, and it is being used in the _parse_array method of the GraphProto class.

anirudhacharya commented on python/mxnet/contrib/onnx/_import/import_onnx.py:
```diff
@@ -0,0 +1,166 @@
+            inputs = [self._nodes[self._renames.get(i, i)] for i in node.input]
+            new_op, mx_attr, inputs = _convert_operator(op_name, onnx_attr, inputs)
+
+            # calling again to get new symbols after some workarounds
```
that line and the corresponding comment is redundant and a mistake. Had put it there for some debugging. Will remove it.

anirudhacharya commented on python/mxnet/contrib/onnx/_import/import_onnx.py:
```diff
@@ -0,0 +1,166 @@
+
+            op = new_op(name=node_name, *inputs, **mx_attr)
+
+            assert len(node.output) == len(op.list_outputs()), (
```
Okay.

anirudhacharya commented:
@piiswrong can you please take a look at the code. We will keep adding more operator support, but can you review the general design of the module. I think I have addressed all your concerns.

CodingCat commented:
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!

marcoabreu commented on tests/python/unittest/test_layers.py:
```diff
@@ -50,5 +50,29 @@ def test_reduce_mean(self):
+    def test_reduce_min(self):
+        """Test for ReduceMin operator"""
+        node_def = helper.make_node("ReduceMin", ["input1"], ["output"], axes=[1, 0], keepdims=1)
+        input1 = self._random_array([3, 10])
```
Please make use of the @with_seed decorator for all tests using randomized data

marcoabreu commented on Jenkinsfile:
```diff
@@ -649,7 +649,20 @@ try {
 
   stage('Integration Test') {
-    parallel 'Python GPU': {
+	parallel 'Python CPU': {
```
Please use the right name. Also, indenting is wrong.

marcoabreu commented on Jenkinsfile:
```diff
@@ -649,7 +649,20 @@ try {
-    parallel 'Python GPU': {
+	parallel 'Python CPU': {
+      node('mxnetlinux-cpu') {
+        ws('workspace/it-python-cpu') {
```
wrong workspace name (should be onnx)

anirudhacharya replied:
will fix this

marcoabreu commented on example/onnx/test_super_resolution.py:
```diff
@@ -0,0 +1,62 @@
+
+download(model_url, 'super_resolution.onnx')
+
+print("Converting onnx format to mxnet's symbol and params...")
```
Could you use logging instead of print?

anirudhacharya replied:
okay. I will make this change.

marcoabreu commented on Jenkinsfile:
```diff
@@ -649,7 +649,20 @@ try {
+          timeout(time: max_time, unit: 'MINUTES') {
+            sh "${docker_run} cpu --dockerbinary docker PYTHONPATH=./python/ python tests/python/integrationtest/onnx_backend_test.py"
+            sh "${docker_run} cpu --dockerbinary docker PYTHONPATH=./python/ python tests/python/integrationtest/test_layers.py"
+            sh "${docker_run} cpu --dockerbinary docker PYTHONPATH=./python/ python example/onnx/test_super_resolution.py"
```
That example does not have any asserts and thus has no effect

marcoabreu commented on example/onnx/test_super_resolution.py:
```diff
@@ -0,0 +1,62 @@
+# specific language governing permissions and limitations
+# under the License.
+
+"""Testing super_resolution model conversion"""
```
Please add asserts if this should be a test

anirudhacharya replied:
it has been added

marcoabreu commented on python/mxnet/contrib/__init__.py:
```diff
@@ -28,5 +28,5 @@
 
 from . import text
-
+from . import onnx
```
What happens if the onnx-package is not present?

anirudhacharya replied:
This is the import for Mxnet's onnx package, which is why I have given the relative path.

marcoabreu replied:
What happens if the onnx-package is not present?

anirudhacharya replied:
This import is for mxnet.contrib.onnx package which will ship as part of the code and not the 'onnx' package which we are installing in CI setup process. So package will always be present.
Following is the folder structure of the new mxnet.contrib.onnx module that is being added in this PR.
mxnet
      ....
      |-- contrib
      |           |-- onnx
      |                     |-- _import
      |-- gluon
                  |-- ...
      ......
Please let me know if this wrong, and I have change something.

marcoabreu replied:
Ah I understand, thanks for elaborating. Could there be a possible namespace clash?

anirudhacharya replied:
No, there is no namespace clash. The import module in onnx folder is name '_import'.

marcoabreu replied:
Ok, thanks

marcoabreu commented on tests/python/integrationtest/test_layers.py:
```diff
@@ -0,0 +1,92 @@
+    Temporary file until we have a corresponding test in onnx-backend_test
+    for these operators."""
+
+    @with_seed(0)
```
Please don't use a default seed. (applies to all with_seeds)

anirudhacharya replied:
will use with_seed() instead of the default 0 value.

marcoabreu commented on tests/python/integrationtest/test_layers.py:
```diff
@@ -0,0 +1,92 @@
+# under the License.
+
+"""Tests for individual operators"""
+# pylint: disable=import-error,no-self-use
```
Please provide reason for disabling pylint here

marcoabreu commented on tests/python/integrationtest/test_layers.py:
```diff
@@ -0,0 +1,92 @@
+# specific language governing permissions and limitations
+# under the License.
+
+"""Tests for individual operators"""
```
Is this test specifically for ONNX? In that case please name it test_onnx.py, otherwise please explain the requirements to run this test

marcoabreu commented on tests/python/integrationtest/onnx_backend_test.py:
```diff
@@ -0,0 +1,102 @@
+# under the License.
+
+"""ONNX test backend wrapper"""
+# pylint: disable=invalid-name,import-error,wrong-import-position
```
Please provide reason to disable Pylint

marcoabreu commented on Jenkinsfile:
```diff
@@ -649,7 +649,20 @@ try {
+          unpack_lib('cpu')
+          timeout(time: max_time, unit: 'MINUTES') {
+            sh "${docker_run} cpu --dockerbinary docker PYTHONPATH=./python/ python tests/python/integrationtest/onnx_backend_test.py"
+            sh "${docker_run} cpu --dockerbinary docker PYTHONPATH=./python/ python tests/python/integrationtest/test_layers.py"
```
Are these two not running nosetest?

anirudhacharya replied:
test_layers is running nosetest and I will change the command, but onnx_backend_test is not using nosetest

marcoabreu replied:
If they're running nosetests, then that's the wrong command to execute them.

For non-nose-tests please create a parallel directory to tests/python because all tests in tests/python are nosetests.

anirudhacharya replied:
onnx_backend_test uses pytest. 

I will shift the whole folder from "tests/python/integrationtest" to "tests/integrationtests". I will run all onnx related tests using pytest. And make corresponding changes in Jenkinsfile. Will that be alright?

marcoabreu replied:
please create a folder called ``tests/python-pytest/onnx`` instead. Right, then you'll have to make the appropriate changes to the dockerfiles

marcoabreu commented on Jenkinsfile:
```diff
@@ -649,7 +649,20 @@ try {
+          init_git()
+          unpack_lib('cpu')
+          timeout(time: max_time, unit: 'MINUTES') {
+            sh "${docker_run} cpu --dockerbinary docker PYTHONPATH=./python/ pytest tests/integrationtests/onnx_backend_test.py"
```
Change path

marcoabreu commented on example/onnx/test_super_resolution.py:
```diff
@@ -0,0 +1,84 @@
+                  img_cb.resize(img_out_y.size, Image.BICUBIC),
+                  img_cr.resize(img_out_y.size, Image.BICUBIC)]).convert("RGB")
+    output_img_dim = 672
+    assert result_img.size == (output_img_dim, output_img_dim)
```
@eric-haibin-lin Does this assert match? I don't think it assesses the task properly.

anirudhacharya replied:
@marcoabreu added more asserts in the import_onnx() method where the actual model import is happening.

marcoabreu replied:
Thanks. But please also verify the output

anirudhacharya replied:
the model is supposed to convert a low resolution image into a high resolution image with specific dimensions, so I have asserted for the size of the result image. What else will I have to assert for?

marcoabreu replied:
I would (don't know if it's feasible) compare the similarity between a simply upscaled picture and the produced picture. If it differs too much, this could indicate that something went wrong and we didn't actually create an upscale picture but just scrambled bits in picture format. 

But when I'm thinking about it... This is an example that is supposed to be used by users as a reference, right? I think we should rather write a test on top of the examples instead of overloading it with verification logic as it could cause issues differentiating the necessary pieces from verification logic.

What do you think?

marcoabreu commented on Jenkinsfile:
```diff
@@ -649,7 +649,20 @@ try {
+          unpack_lib('cpu')
+          timeout(time: max_time, unit: 'MINUTES') {
+            sh "${docker_run} cpu --dockerbinary docker PYTHONPATH=./python/ pytest tests/python-pytest/onnx/onnx_backend_test.py"
+            sh "${docker_run} cpu --dockerbinary docker PYTHONPATH=./python/ pytest tests/python-pytest/onnx/test_onnx.py"
```
inconsistent filenames

anirudhacharya replied:
@marcoabreu do you mean that either both should be ("test_onnx" and "test_onnx_backend") or ("onnx_backend_test" and "onnx_test")?

marcoabreu commented on tests/python-pytest/onnx/test_onnx.py:
```diff
@@ -0,0 +1,86 @@
+from onnx import helper
+import backend as mxnet_backend
+CURR_PATH = os.path.dirname(os.path.abspath(os.path.expanduser(__file__)))
+sys.path.insert(0, os.path.join(CURR_PATH, '../../python/unittest'))
```
What is this for?

anirudhacharya replied:
It is for the withseed decorator.

marcoabreu commented on example/onnx/test_super_resolution.py:
```diff
@@ -0,0 +1,112 @@
+def import_onnx():
+    """Import the onnx model into mxnet"""
+    LOGGER.info("Converting onnx format to mxnet's symbol and params...")
+    sym, params = onnx_mxnet.import_model('super_resolution.onnx')
```
Is the version (or hash) of this file verified somewhere? To prevent it being updated without us noticing

anirudhacharya replied:
No, it currently does not. I will add the verification. Can you please point me to some examples where this is done, for reference.

marcoabreu replied:
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/data/vision/datasets.py#L152

We're not adopting this behaviour everywhere, but it would be good if we could :)

anirudhacharya replied:
This PR currently has tests for all the operator translations, I plan on making a follow up PR for adding more tests for onnx model translations. I will add a version to the above files and make this change in that follow up PR. Will that be okay?

marcoabreu replied:
I would prefer tests being introduced at the same time as a new feature. How much is missing?

marcoabreu commented on Jenkinsfile:
```diff
@@ -649,7 +649,20 @@ try {
 
   stage('Integration Test') {
-    parallel 'Python GPU': {
+  	parallel 'Onnx CPU': {
```
Indentation

marcoabreu commented on tests/python-pytest/onnx/test_onnx.py:
```diff
@@ -0,0 +1,86 @@
+# specific language governing permissions and limitations
+# under the License.
+
+"""Tests for individual operators"""
```
This is not using the onnx tests, right? In that case, please move them to the other folder and use nosetests since that framework is the default for now

anirudhacharya replied:
I will move it.
This file actually contains tests which currently don't exist on the ONNX repo. Once we make a couple of PRs on ONNX repo and they get merged, this file will get EOLed.

marcoabreu replied:
Ah I see. If the plan is to integrate them into ONNX, then it's fine. Please just consider nosetests for tests that are specifically for the MXNet side

anirudhacharya commented on tests/ci_build/install/ubuntu_install_onnx.sh:
```diff
@@ -0,0 +1,26 @@
+apt-get install -y libprotobuf-dev protobuf-compiler
+
+pip2 install pytest==3.4.0 pytest-cov==2.5.1 protobuf==3.0.0 onnx==1.0.1 Pillow==5.0.0 tabulate==0.7.5
+pip3 install pytest==3.4.0 pytest-cov==2.5.1 protobuf==3.0.0 onnx==1.0.1 Pillow==5.0.0 tabulate==0.7.5
```
@marcoabreu in this file instead of pip installing onnx version 1.0.1 would it be okay to do the following  -
```bash
git clone --recursive https://github.com/onnx/onnx.git
cd onnx
git checkout 7e205b66190f4376c64741ba7705dc23e9fbf225
python setup.py install
```
This will allow us to pull in more recent changes from the onnx repo, which will be needed for running a lot of our tests. For example, if we add a new operator later then we will also add corresponding tests on the onnx repo which will need to be run on the CI pipeline.  And we cannot always rely on onnx pip package release, which could be outdated. Even in the current CI run, some tests are probably failing because the onnx 1.0.1 release is not up to date.
Would this change be okay in this script or do you have any other suggestions?

marcoabreu replied:
Yes, it's fine as long as a commit id or version tag is set. Ideally, we'd use a stable tag instead of a commit hash, but considering the recent and active development as well as number of changes at ONNX, it's okay. It would be good to have a stable version for reference because this will allow us that we verified compatibility between MXNet Version 1.X and ONNX 1.X.

marcoabreu commented on Jenkinsfile:
```diff
@@ -560,7 +560,18 @@ try {
 
   stage('Integration Test') {
-    parallel 'Python GPU': {
+  parallel 'Onnx CPU': {
```
Indenting

marcoabreu commented on ci/docker/Dockerfile.build.ubuntu_cpu:
```diff
@@ -46,6 +46,8 @@ COPY install/ubuntu_docs.sh /work/
 RUN /work/ubuntu_docs.sh
 COPY install/ubuntu_adduser.sh /work/
 RUN /work/ubuntu_adduser.sh
+COPY install/ubuntu_onnx.sh /work/
```
Please put this script before 'adduser'

marcoabreu commented on tests/python-pytest/onnx/onnx_backend_test.py:
```diff
@@ -0,0 +1,116 @@
+
+BACKEND_TEST = onnx.backend.test.BackendTest(mxnet_backend, __name__)
+
+IMPLEMENTED_OPERATORS = [
```
Are you going to activate all tests?

marcoabreu commented on tests/python-pytest/onnx/onnx_test.py:
```diff
@@ -0,0 +1,138 @@
+# set up logger
+logging.basicConfig()
+LOGGER = logging.getLogger()
+LOGGER.setLevel(logging.INFO)
```
Please remove this

marcoabreu commented on tests/python-pytest/onnx/onnx_test.py:
```diff
@@ -0,0 +1,138 @@
+    output = mxnet_backend.run_node(node_def, [input1])[0]
+    npt.assert_almost_equal(output, np.squeeze(input1, axis=[1, 3]))
+
+def test_super_resolution():
```
test_super_resolution_example

marcoabreu commented on tests/python-pytest/onnx/onnx_test.py:
```diff
@@ -0,0 +1,133 @@
+                                    'param_1', 'param_0', 'param_3', 'param_2']):
+        assert param_item in param_keys
+
+    logging.getLogger().info("Asserted the result of the onnx model conversion")
```
logging.info; getLogger() is not necessary

marcoabreu commented on tests/python-pytest/onnx/onnx_test.py:
```diff
@@ -0,0 +1,133 @@
+    input_image, img_cb, img_cr = super_resolution.get_test_image()
+    result_img = super_resolution.perform_inference(sym, params, input_image,
+                                                    img_cb, img_cr)
+    assert result_img.size == (output_img_dim, output_img_dim)
```
Please add a correctness test. Hashing the output and comparing that could be a good way.

Roshrini commented on python/mxnet/contrib/onnx/_import/import_helper.py:
```diff
@@ -0,0 +1,105 @@
+from .op_translations import reduce_prod, avg_pooling, max_pooling
+from .op_translations import argmax, argmin, maximum, minimum
+
+# convert_map defines maps of ONNX operator names to converter functor(callable)
```
converter functor? typo?

anirudhacharya replied:
its a typo in the comment. I will fix it.

Roshrini commented on python/mxnet/contrib/onnx/_import/import_model.py:
```diff
@@ -0,0 +1,46 @@
+    try:
+        import onnx
+    except ImportError:
+        raise ImportError("Onnx and protobuf need to be installed")
```
Can we also add a link to onnx installation here while throwing this error?

Roshrini commented on python/mxnet/contrib/onnx/_import/import_model.py:
```diff
@@ -0,0 +1,46 @@
+    ----------
+    model_file : ONNX model file name
+
+    :returns (sym, params)
```
Check for doc rendering. Will be helpful if we decide to add these API docs on the website later.
Follow the same convention throughout

Roshrini commented on python/mxnet/contrib/onnx/_import/import_helper.py:
```diff
@@ -0,0 +1,105 @@
+from .op_translations import leaky_relu, _elu, _prelu, softmax, fully_connected
+from .op_translations import global_avgpooling, global_maxpooling, linalg_gemm
+from .op_translations import sigmoid, pad, relu, matrix_multiplication, batch_norm
+from .op_translations import dropout, local_response_norm, conv, deconv
```
why not import all functions from .op_translations instead of specifying each function here?

anirudhacharya replied:
it is not as per mxnet convention to do 
```python
from op_translations import *
```
It was an earlier comment by one of the reviewers/committers.

nswamy commented on python/mxnet/test_utils.py:
```diff
@@ -1367,7 +1367,8 @@ def list_gpus():
 
-def download(url, fname=None, dirname=None, overwrite=False):
+
+def download(url, fname=None, dirname=None, overwrite=False, version_tag=None):
```
rename `version_tag` to `eTag`, lets not define a new concept

nswamy replied:
Can you make a separate PR for this change?

nswamy commented on python/mxnet/test_utils.py:
```diff
@@ -1407,12 +1410,15 @@ def download(url, fname=None, dirname=None, overwrite=False):
                     raise OSError('failed to create ' + dirname)
 
-    if not overwrite and os.path.exists(fname):
+    if not overwrite and os.path.exists(fname) and not version_tag:
```
how do you plan to handle if i give a different eTag

anirudhacharya replied:
Have handled it in line 1419-1420

nswamy commented on python/mxnet/test_utils.py:
```diff
@@ -1407,12 +1410,15 @@ def download(url, fname=None, dirname=None, overwrite=False):
     assert r.status_code == 200, "failed to open %s" % url
+    if version_tag and r.headers['ETag'] != version_tag:
+        logging.info("The version tag of the file does not match the expected version. "
+                     + "Proceeding with the file download...")
```
why do you have to download if they don't match

anirudhacharya replied:
the download method was already present in the test_utils module and in use by other modules in the project. I did not want the eTag mismatch to fail any existing tests.
Also I used this as a reference - https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/utils.py#L216-L220

marcoabreu replied:
Oh right, good point. We should rather fail if the file is not the one we are expecting.

reminisce commented:
Is there a plan to add a doc page with these changes? Normally we don't expect users to find out APIs by reading the code.

anirudhacharya commented:
@reminisce there is a follow up PR with the doc page changes.
